File: Serial_cmd_gui.py
#
## Copyright (c) 2021, Bradley A. Minch
## All rights reserved.
##
## Redistribution and use in source and binary forms, with or without
## modification, are permitted provided that the following conditions are met: 
## 
##     1. Redistributions of source code must retain the above copyright 
##        notice, this list of conditions and the following disclaimer. 
##     2. Redistributions in binary form must reproduce the above copyright 
##        notice, this list of conditions and the following disclaimer in the 
##        documentation and/or other materials provided with the distribution. 
##
## THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
## AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
## IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
## ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
## LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
## CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
## SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
## INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
## CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
## POSSIBILITY OF SUCH DAMAGE.
#

import tkinter as tk
import Serial_cmd

class Serial_cmd_gui:

    def __init__(self):
        self.dev = Serial_cmd.Serial_cmd()
        if self.dev.dev is not None:
            self.update_job = None
            self.root = tk.Tk()
            self.root.title('Serial_cmd GUI')
            self.root.protocol('WM_DELETE_WINDOW', self.shut_down)
            fm = tk.Frame(self.root)
            tk.Button(fm, text = 'TOGGLE RED', command = self.toggle_red).pack(side = tk.LEFT)
            fm.pack(side = tk.TOP)
            green_slider = tk.Scale(self.root, from_ = 0, to = 1000, orient = tk.HORIZONTAL, showvalue = tk.FALSE, command = self.set_green_callback)
            green_slider.set(500)
            green_slider.pack(side = tk.TOP)
            yellow_slider = tk.Scale(self.root, from_ = 0, to = 1000, orient = tk.HORIZONTAL, showvalue = tk.FALSE, command = self.set_yellow_callback)
            yellow_slider.set(500)
            yellow_slider.pack(side = tk.TOP)
            self.sw1_status = tk.Label(self.root, text = 'SW1 is currently ?')
            self.sw1_status.pack(side = tk.TOP)
            self.sw2_status = tk.Label(self.root, text = 'SW2 is currently ?')
            self.sw2_status.pack(side = tk.TOP)
            self.pot_status = tk.Label(self.root, text = 'POT is currently ????')
            self.pot_status.pack(side = tk.TOP)
            self.update_job = self.root.after(500, self.update_status)

    def toggle_red(self):
        self.dev.set_red(1 - self.dev.get_red())

    def set_green_callback(self, value):
        self.dev.set_green(int(value))

    def set_yellow_callback(self, value):
        self.dev.set_yellow(int(value))

    def update_status(self):
        self.sw1_status.configure(text = 'SW1 is currently {!s}'.format(self.dev.get_sw1()))
        self.sw2_status.configure(text = 'SW2 is currently {!s}'.format(self.dev.get_sw2()))
        self.pot_status.configure(text = 'POT is currently {:04d}'.format(self.dev.get_pot()))
        self.update_job = self.root.after(50, self.update_status)

    def shut_down(self):
        if self.update_job is not None:
            self.root.after_cancel(self.update_job)
        self.root.destroy()

if __name__=='__main__':
    gui = Serial_cmd_gui()
    gui.root.mainloop()

File: Serial_cmd_gui.m
%%
%% Copyright (c) 2021, Bradley A. Minch
%% All rights reserved.
%%
%% Redistribution and use in source and binary forms, with or without
%% modification, are permitted provided that the following conditions are met: 
%% 
%%     1. Redistributions of source code must retain the above copyright 
%%        notice, this list of conditions and the following disclaimer. 
%%     2. Redistributions in binary form must reproduce the above copyright 
%%        notice, this list of conditions and the following disclaimer in the 
%%        documentation and/or other materials provided with the distribution. 
%%
%% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
%% AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
%% IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
%% ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
%% LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
%% CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
%% SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
%% INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
%% CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
%% ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
%% POSSIBILITY OF SUCH DAMAGE.
%%

function Serial_cmd_gui(dev)

dev_name = inputname(1);

fig_handles = get(0, 'children');
dev_gui_fig = findobj(fig_handles, 'tag', [dev_name, '_gui_fig']);
if ~isempty(dev_gui_fig)
    figure(dev_gui_fig);
    return;
end

set(0, 'units', 'pixels');
screen_rect = get(0, 'screensize');
screen_width = screen_rect(3); 
screen_height = screen_rect(4);

fig = figure;
fig.Units = 'pixels';
fig.Position = [(screen_width / 2 - 200), (screen_height / 2 - 90), 400, 180];
fig.NumberTitle = 'off';
fig.Name = ['Serial_cmd GUI (', dev_name, ')'];
fig.Resize = 'off';
fig.Tag = [dev_name, '_gui_fig'];
fig.DeleteFcn = @stop_update_timer;

red_btn = uicontrol(fig);
red_btn.Style = 'pushbutton';
red_btn.String = 'TOGGLE RED';
red_btn.Position = [160, 150, 80, 24];
red_btn.Callback = @toggle_red_callback;

    function toggle_red_callback(src, event)
        dev.set_red(uint16(1 - dev.get_red()));
    end

green_slider = uicontrol(fig);
green_slider.Style = 'slider';
green_slider.Min = 0;
green_slider.Max = 1000;
green_slider.Value = dev.get_green();
green_slider.SliderStep = [1, 10];
green_slider.Position = [50, 120, 300, 24];
green_slider.Callback = @green_slider_callback;

    function green_slider_callback(src, event)
        dev.set_green(uint16(src.Value));
    end

yellow_slider = uicontrol(fig);
yellow_slider.Style = 'slider';
yellow_slider.Min = 0;
yellow_slider.Max = 1000;
yellow_slider.Value = dev.get_yellow();
yellow_slider.SliderStep = [1, 10];
yellow_slider.Position = [50, 90, 300, 24];
yellow_slider.Callback = @yellow_slider_callback;

    function yellow_slider_callback(src, event)
        dev.set_yellow(uint16(src.Value));
    end

sw1_text = uicontrol(fig);
sw1_text.Style = 'text';
sw1_text.Position = [50, 60, 300, 24];
sw1_text.String = 'SW1 is currently ?';

sw2_text = uicontrol(fig);
sw2_text.Style = 'text';
sw2_text.Position = [50, 30, 300, 24];
sw2_text.String = 'SW2 is currently ?';

pot_text = uicontrol(fig);
pot_text.Style = 'text';
pot_text.Position = [50, 0, 300, 24];
pot_text.String = 'POT is currently ????';

update_timer = timer;
update_timer.ExecutionMode = 'fixedSpacing';
update_timer.Period = 0.05;
update_timer.StartDelay = 0;
update_timer.TimerFcn = @update_status;
start(update_timer);

    function update_status(src, event)
        sw1_text.String = sprintf('SW1 is currently %d', dev.get_sw1());
        sw2_text.String = sprintf('SW2 is currently %d', dev.get_sw2());
        pot_text.String = sprintf('POT is currently %04d', dev.get_pot());
    end

    function stop_update_timer(src, event)
        stop(update_timer);
        delete(update_timer);
    end

end
